File: apps/api-rs/src/api/system_tenant_invitations.rs
use axum::extract::{Path, State};
use axum::Json;
use chrono::{Duration, Utc};
use serde::Deserialize;
use serde_json::json;
use sha2::{Digest, Sha256};
use sqlx::Row;
use uuid::Uuid;

use crate::auth::{require_super_admin, ActorExtractor};
use crate::error::AppError;
use crate::state::AppState;

#[derive(Debug, Deserialize)]
pub struct GenerateAdminInvitationRequest {
    expires_in_days: Option<i64>,
}

pub async fn generate_admin_invitation(
    State(state): State<AppState>,
    ActorExtractor(actor): ActorExtractor,
    Path(tenant_id): Path<Uuid>,
    Json(req): Json<GenerateAdminInvitationRequest>,
) -> Result<Json<serde_json::Value>, AppError> {
    require_super_admin(&actor)?;
    let pool = state
        .db_pool
        .as_ref()
        .ok_or_else(|| AppError::bad_request("DB_REQUIRED", "租户管理功能需要数据库"))?;
    let tenant_status: String = sqlx::query_scalar("SELECT status FROM tenant WHERE id = $1")
        .bind(tenant_id)
        .fetch_optional(pool)
        .await?
        .ok_or_else(|| AppError::NotFound {
            code: "TENANT_NOT_FOUND".to_string(),
            message: "租户不存在".to_string(),
        })?;
    if !matches!(tenant_status.as_str(), "pending" | "active") {
        return Err(AppError::Conflict {
            code: "TENANT_NOT_INVITABLE".to_string(),
            message: "只有待加入或运行中的租户可以邀请管理员".to_string(),
        });
    }

    let token = format!("inv_{}{}", Uuid::new_v4().simple(), Uuid::new_v4().simple());
    let token_hash = hex::encode(Sha256::digest(token.as_bytes()));
    let expires_at = Utc::now() + Duration::days(req.expires_in_days.unwrap_or(7).clamp(1, 30));
    let mut transaction = pool.begin().await?;
    let row = sqlx::query(
        r#"
        INSERT INTO tenant_invitation
          (tenant_id, email, name, roles, kb_grants, token_hash, status, invited_by, expires_at)
        VALUES ($1, NULL, NULL, ARRAY['tenant_admin'], '[]'::jsonb, $2, 'pending', $3, $4)
        ON CONFLICT (tenant_id) WHERE status = 'pending' AND email IS NULL
        DO UPDATE SET token_hash = EXCLUDED.token_hash,
                      expires_at = EXCLUDED.expires_at,
                      revoked_at = NULL,
                      updated_at = NOW()
        RETURNING id, expires_at
        "#,
    )
    .bind(tenant_id)
    .bind(&token_hash)
    .bind(actor.user_id)
    .bind(expires_at)
    .fetch_one(&mut *transaction)
    .await?;
    let invitation_id: Uuid = row.get("id");
    sqlx::query(
        r#"
        INSERT INTO audit_log
          (tenant_id, actor_user_id, actor_role, action, resource_type, resource_id, detail)
        VALUES ($1, $2, 'super_admin', 'tenant.admin_invitation.generate',
                'tenant_invitation', $3, $4)
        "#,
    )
    .bind(tenant_id)
    .bind(actor.user_id)
    .bind(invitation_id.to_string())
    .bind(json!({
        "expires_at": row.get::<chrono::DateTime<Utc>, _>("expires_at"),
    }))
    .execute(&mut *transaction)
    .await?;
    transaction.commit().await?;

    Ok(Json(json!({
        "id": invitation_id,
        "expires_at": row.get::<chrono::DateTime<Utc>, _>("expires_at"),
        "invite_url": format!("/invite?token={token}"),
    })))
}

File: apps/api-bun/src/api/system_tenant_invitations.ts
import { Hono } from 'hono';
import type { Context } from 'hono';
import type { Sql } from 'postgres';
import { AppError } from '../errors.ts';
import type { AppEnv } from '../http/types.ts';
import type { AppState } from '../state.ts';
import { requireSuperAdmin } from '../auth/permissions.ts';
import { toRfc3339 } from '../infra/time.ts';
import { clampDays, invitationTokenHash, newInvitationToken } from './system_tenants.ts';

interface OwnerInvitationRequest { expires_in_days?: number | null; }

export function systemTenantInvitationsRouter(): Hono<AppEnv> {
  const router = new Hono<AppEnv>();
  router.post('/api/v1/system/tenants/:id/owner-invitation', createOwnerInvitation);
  router.post('/api/v1/system/tenants/:id/owner-invitation/resend', resendOwnerInvitation);
  router.post('/api/v1/system/tenants/:id/owner-invitation/revoke', revokeOwnerInvitation);
  return router;
}

async function createOwnerInvitation(c: Context<AppEnv>) {
  const { sql, actor, tenantId } = invitationContext(c);
  await ensureTenantCanReceiveOwner(sql, tenantId);
  const pending = await sql`
    SELECT 1 FROM tenant_invitation
    WHERE tenant_id = ${tenantId} AND kind = 'bootstrap_owner' AND status = 'pending'
    LIMIT 1
  `;
  if (pending[0]) {
    throw AppError.conflictWith('OWNER_INVITATION_ALREADY_PENDING', '租户已有待接受的所有者邀请');
  }
  const req = await optionalJson<OwnerInvitationRequest>(c);
  const { token, hash, expiresAt } = newToken(req.expires_in_days);
  const rows = await sql`
    INSERT INTO tenant_invitation (
      tenant_id, token_hash, kind, invitee_username_normalized, roles,
      kb_grants, status, invited_by, expires_at
    )
    VALUES (
      ${tenantId}, ${hash}, 'bootstrap_owner', NULL, ARRAY['tenant_owner'],
      '[]'::jsonb, 'pending', ${actor.user_id}, ${expiresAt}
    )
    RETURNING id, expires_at
  `;
  await sql`
    INSERT INTO audit_log
      (tenant_id, actor_user_id, actor_role, action, resource_type, resource_id, detail)
    VALUES (
      ${tenantId}, ${actor.user_id}, 'super_admin', 'tenant.owner_invitation.create',
      'tenant_invitation', ${String(rows[0]!.id)}, '{}'::jsonb
    )
  `;
  return c.json(ownerInvitation(rows[0]!, token), 201);
}

async function resendOwnerInvitation(c: Context<AppEnv>) {
  const { sql, actor, tenantId } = invitationContext(c);
  await ensureTenantCanReceiveOwner(sql, tenantId);
  const req = await optionalJson<OwnerInvitationRequest>(c);
  const { token, hash, expiresAt } = newToken(req.expires_in_days);
  const rows = await sql`
    UPDATE tenant_invitation
    SET token_hash = ${hash}, expires_at = ${expiresAt},
        revoked_at = NULL, updated_at = NOW()
    WHERE tenant_id = ${tenantId} AND kind = 'bootstrap_owner' AND status = 'pending'
    RETURNING id, expires_at
  `;
  if (!rows[0]) {
    throw AppError.notFound('OWNER_INVITATION_NOT_FOUND', '待接受的租户所有者邀请不存在');
  }
  await sql`
    INSERT INTO audit_log
      (tenant_id, actor_user_id, actor_role, action, resource_type, resource_id, detail)
    VALUES (
      ${tenantId}, ${actor.user_id}, 'super_admin', 'tenant.owner_invitation.resend',
      'tenant_invitation', ${String(rows[0].id)}, '{}'::jsonb
    )
  `;
  return c.json(ownerInvitation(rows[0], token));
}

async function revokeOwnerInvitation(c: Context<AppEnv>) {
  const { sql, actor, tenantId } = invitationContext(c);
  const rows = await sql`
    UPDATE tenant_invitation
    SET status = 'revoked', revoked_at = NOW(), updated_at = NOW()
    WHERE tenant_id = ${tenantId} AND kind = 'bootstrap_owner' AND status = 'pending'
    RETURNING id
  `;
  if (!rows[0]) {
    throw AppError.notFound('OWNER_INVITATION_NOT_FOUND', '待接受的租户所有者邀请不存在');
  }
  await sql`
    INSERT INTO audit_log
      (tenant_id, actor_user_id, actor_role, action, resource_type, resource_id, detail)
    VALUES (
      ${tenantId}, ${actor.user_id}, 'super_admin', 'tenant.owner_invitation.revoke',
      'tenant_invitation', ${String(rows[0].id)}, '{}'::jsonb
    )
  `;
  return c.json({ invitation: { id: String(rows[0].id), status: 'revoked' } });
}

function invitationContext(c: Context<AppEnv>) {
  const state = c.get('appState');
  const actor = c.get('actor');
  requireSuperAdmin(actor);
  return { sql: requiredSql(state), actor, tenantId: c.req.param('id')! };
}

async function ensureTenantCanReceiveOwner(sql: Sql, tenantId: string): Promise<void> {
  const tenants = await sql`
    SELECT status FROM tenant WHERE id = ${tenantId}
  `;
  if (!tenants[0]) throw AppError.notFound('TENANT_NOT_FOUND', '租户不存在');
  if (!['pending', 'active'].includes(String(tenants[0].status))) {
    throw AppError.conflictWith('TENANT_NOT_INVITABLE', '租户当前不可邀请所有者');
  }
  const owners = await sql`
    SELECT 1 FROM tenant_member
    WHERE tenant_id = ${tenantId} AND status = 'active'
      AND 'tenant_owner' = ANY(roles)
    LIMIT 1
  `;
  if (owners[0]) throw AppError.conflictWith('TENANT_OWNER_EXISTS', '租户已有所有者');
}

function newToken(expiresInDays: number | null | undefined) {
  const token = newInvitationToken();
  return {
    token,
    hash: invitationTokenHash(token),
    expiresAt: new Date(Date.now() + clampDays(expiresInDays ?? 7) * 86_400_000),
  };
}

function ownerInvitation(row: Record<string, unknown>, token: string) {
  return {
    invitation: {
      id: String(row.id),
      kind: 'bootstrap_owner',
      invitee_username: null,
      roles: ['tenant_owner'],
      status: 'pending',
      expires_at: toRfc3339(new Date(row.expires_at as Date | string)),
    },
    invite_url: `/invite#token=${encodeURIComponent(token)}`,
  };
}

async function optionalJson<T>(c: Context<AppEnv>): Promise<Partial<T>> {
  const text = await c.req.text();
  return text ? JSON.parse(text) as Partial<T> : {};
}

function requiredSql(state: AppState): Sql {
  if (!state.sql) throw AppError.badRequest('DB_REQUIRED', '租户管理功能需要数据库');
  return state.sql;
}
